### PR TITLE
GCP: Container Clusters

### PR DESCRIPTION
**Pull request by rambleraptor**
##### SUMMARY
Adding support for GCP (Kubernetes Engine) Container Clusters 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
gcp_container_cluster

##### ANSIBLE VERSION
```
2.6
```

##### ADDITIONAL INFORMATION
This was autogenerated using [Magic Modules](https://github.com/GoogleCloudPlatform/magic-modules)

### CODE REVIEW

File: lib/ansible/modules/cloud/google/gcp_container_cluster.py [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+    module.exit_json(**fetch)
+
+
+def create(module, link):
+    auth = GcpSession(module, 'container')
+    return wait_for_operation(module, auth.post(link, resource_to_request(module)))
+
+
+def update(module, link, fetch):
+    module.fail_json(msg="Cluster cannot be edited")
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_container_cluster.py]: How are edits supposed to be performed?

File: lib/ansible/modules/cloud/google/gcp_container_cluster.py [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+    wait_done = wait_for_completion(status, op_result, module)
+    return fetch_resource(module, navigate_hash(wait_done, ['targetLink']))
+
+
+def wait_for_completion(status, op_result, module):
+    op_id = navigate_hash(op_result, ['name'])
+    op_uri = async_op_url(module, {'op_id': op_id})
+    while status != 'DONE':
+        raise_if_errors(op_result, ['error', 'errors'], 'message')
+        time.sleep(1.0)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_container_cluster.py:799]: What rate limits are there on this API? Any account-wide ones will cause inconsistent failures if many playbooks are run at once. On the AWS modules we have common exponential-backoff code that handles this, and the base class could be extended for GCP.

**Review by ryansb — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/google/gcp_container_cluster.py (hunk 2 above)
Comment (rambleraptor) [reply lib/ansible/modules/cloud/google/gcp_container_cluster.py:799]: I haven't been able to find a good answer for the rate limits on this API. I'm not worried about the potential for it to fail though. We can do some kind of exponential-backoff, but I'm thinking that might be better as an additional feature later on

File: lib/ansible/modules/cloud/google/gcp_container_cluster.py [hunk 3]
```diff
@@ -0,0 +1,1036 @@
...
+                description:
+                    - The Google Cloud Platform Service Account to be used by the
+                      node VMs.  If no Service Account is specified, the "default"
+                      service account is used.
+                required: false
+            metadata:
+                description:
+                    - The metadata key/value pairs assigned to instances in the
+                      cluster.
+                    - |
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_container_cluster.py]: This seems to only happen on second-paragraph and beyond rendering, if that helps you track it down.

**Review by ryansb — CHANGES_REQUESTED** [review]

**Review by s-hertel — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Pull request merged**